devmand/gateway/src/devmand/channels/cli/SshSessionAsync.cpp

```diff
@@ -66,9 +66,12 @@ Future<Unit> SshSessionAsync::write(const string& command) {
 }
 
 Future<string> SshSessionAsync::readUntilOutput(const string& lastOutput) {
-  return via(executor.get(), [this, lastOutput] {
-    return this->readUntilOutputBlocking(lastOutput);
-  });
+  this->readingState.currentLastOutput = lastOutput;
+  this->readingState.promise = std::make_shared<Promise<string>>();
+  this->readingState.outputSoFar = "";
```

Comment:
do we need both the queue and outputSoFar ? Isn't that redundant ?

Reply:
I was also thinking about it - and yes it is redundant. 

The reason for pop-ing the queue and storing it as "outputSoFar" is because we need to perform string checks on the content. If we wouldn't keep it as string we would have to pop the whole queue concatenate it to string, make the string check and if it fails throw the string away. Not to mention we would need to somehow get the pop-ed substrings back to the queue, but since it's a single producer it would be difficult.

The solution where we pop the content  and store is as a persistent string seems more elegant.